lib/diff/lcs.rb

```diff
@@ -292,37 +292,32 @@ def traverse_sequences(seq1, seq2, callbacks = Diff::LCS::SequenceCallbacks) #:y
     b_size = seq2.size
     ai = bj = 0
 
-    (0...matches.size).each do |i|
```

Comment:
`i` will be increased by `1` after each iteration, in the new code, `ai` will also be increased by `1` after each iteration, so they're the same, and we can ax `i` so that we have one less variable to keep track of.

Reply:
I think that there’s a couple of minor performance improvements we can do:

```diff --git c/lib/diff/lcs.rb i/lib/diff/lcs.rb
index f3716e1c5378..64f46c957de3 100644
--- c/lib/diff/lcs.rb
+++ i/lib/diff/lcs.rb
@@ -293,16 +293,18 @@ def traverse_sequences(seq1, seq2, callbacks = Diff::LCS::SequenceCallbacks) #:y
     ai = bj = 0
 
     matches.each do |b_line|
-      ax = string ? seq1[ai, 1] : seq1[ai]
-      bx = string ? seq2[bj, 1] : seq2[bj]
-
       if b_line.nil?
-        unless ax.nil? or (string and ax.empty?)
+        unless seq1[ai].nil?
+          ax = string ? seq1[ai, 1] : seq1[ai]
+          bx = string ? seq2[bj, 1] : seq2[bj]
+
           event = Diff::LCS::ContextChange.new('-', ai, ax, bj, bx)
           event = yield event if block_given?
           callbacks.discard_a(event)
         end
       else
+        ax = string ? seq1[ai, 1] : seq1[ai]
+
         loop do
           break unless bj < b_line
 
@@ -312,10 +314,13 @@ def traverse_sequences(seq1, seq2, callbacks = Diff::LCS::SequenceCallbacks) #:y
           callbacks.discard_b(event)
           bj += 1
         end
+
         bx = string ? seq2[bj, 1] : seq2[bj]
+
         event = Diff::LCS::ContextChange.new('=', ai, ax, bj, bx)
         event = yield event if block_given?
         callbacks.match(event)
+
         bj += 1
       end
       ai += 1
```

Reply:
This simplifies the test from `unless ax.nil? or (string and ax.empty?)` and defers setting `ax` and `bx` until they are needed, which is particularly important for the `else` condition because `bx` is immediately reset.